core/tests/test_handlers.py

```diff
@@ -556,3 +558,101 @@ class TestAsyncResult(AsyncResult):
         match=r"Can't check Test result directly; use await or an on_result handler",
     ):
         result != 42
+
+
+def test_async_result(event_loop):
+    """An async result can be set"""
+    on_result = Mock()
+
+    result = ExampleAsyncResult(on_result)
+
+    result.set_result(42)
+
+    # Evaluate the future
+    async_answer = event_loop.run_until_complete(result.future)
+
+    # The answer was returned, and passed to the callback
+    assert async_answer == 42
+    on_result.assert_called_once_with(42)
+
+
+def test_async_result_cancelled(event_loop):
+    """An async result can be set even if the future is cancelled"""
+    on_result = Mock()
+
+    result = ExampleAsyncResult(on_result)
+
+    # cancel the future.
+    result.future.cancel()
+
+    result.set_result(42)
+
+    # Evaluate the future. This will raise an error
+    with pytest.raises(asyncio.CancelledError):
+        event_loop.run_until_complete(result.future)
+
+    # The answer was still passed to the callback
+    on_result.assert_called_once_with(42)
```

Comment:
The original Android code didn't do this, on the basis that since we're exposing a cancellation mechanism, it would make sense for that to cancel the `on_result` call as well, because the app probably isn't able to accept it after cancellation.

Reply:
Makes sense; I've updated the code to do this.